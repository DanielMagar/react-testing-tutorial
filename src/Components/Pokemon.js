export const getPokemon = () => {
  let name = "charizard"
  return fetch(`https://pokeapi.co/api/v2/pokemon/${name}/`)
    .then((response) => {
      if (!response.ok) {
        throw new Error(`Network response was not ok: ${response.statusText}`);
      }
      return response.json(); // Return the result of response.json()
    })
    .catch(error => {
      throw new Error('Failed to fetch data', error); // Handle errors here
    });
};
