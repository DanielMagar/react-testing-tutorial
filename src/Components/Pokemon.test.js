// pokemonApi.test.js
import { getPokemon } from "./Pokemon";

// Mock the global fetch function
global.fetch = jest.fn();

describe('getPokemon', () => {
  afterEach(() => {
    jest.clearAllMocks(); // Clear mocks after each test
  });

  it('should fetch PokÃ©mon data successfully', async () => {
    const mockData = {
      name: 'charizard',
      abilities: [
        { ability: { name: 'blaze' } },
        { ability: { name: 'solar-power' } },
      ],
    };

    // Mocking the fetch response
    fetch.mockResolvedValueOnce({
      ok: true,
      json: jest.fn().mockResolvedValueOnce(mockData), // Mock the json method
    });

    const data = await getPokemon('charizard');

    expect(fetch).toHaveBeenCalledWith('https://pokeapi.co/api/v2/pokemon/charizard/');
    expect(data).toEqual(mockData); // Check if returned data matches mock data
    expect(data).toMatchSnapshot();
  });

  it('should handle errors when fetch fails', async () => {
    // Mocking the fetch to throw an error
    fetch.mockRejectedValueOnce(new Error('Network Error'));

    await expect(getPokemon('unknown')).rejects.toThrow('Failed to fetch data');
  });
  it('should handle non-ok response status', async () => {
    // Mocking fetch to return an ok: false response (e.g., 404)
    fetch.mockResolvedValueOnce({
      ok: false,
      json: jest.fn(), // Mock the json method (won't be called)
    });

    await expect(getPokemon('unknown')).rejects.toThrow('Failed to fetch data');
    expect(fetch).toHaveBeenCalledWith('https://pokeapi.co/api/v2/pokemon/charizard/');
  });
  it('should handle missing abilities in the response', async () => {
    const mockData = {
      name: 'charizard',
      abilities: [], // No abilities
    };

    fetch.mockResolvedValueOnce({
      ok: true,
      json: jest.fn().mockResolvedValueOnce(mockData), // Mock the json method
    });

    const data = await getPokemon('charizard');

    expect(data.abilities).toEqual([]); // Should handle empty abilities list gracefully
  });
});
