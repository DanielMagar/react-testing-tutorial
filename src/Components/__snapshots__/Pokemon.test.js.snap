// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPokemon should fetch Pok√©mon data successfully 1`] = `
Object {
  "abilities": Array [
    Object {
      "ability": Object {
        "name": "blaze",
      },
    },
    Object {
      "ability": Object {
        "name": "solar-power",
      },
    },
  ],
  "name": "charizard",
}
`;
